File: src/components/asteroid.py

import math

import pygame as pg
from pygame import Vector2

from .. import util
from .space_object import SpaceObject

BIG = 32
MEDIUM = 16
SMALL = 8

YELLOW = (255, 255, 0)
BLACK = (0, 0, 0)

class Asteroid(SpaceObject):

	last_asteroid_id = 0
	
	@classmethod
	def get_next_asteroid_id(cls):
		result = cls.last_asteroid_id
		cls.last_asteroid_id += 1
		return result

	def __init__(self, x, y, size, rng):
		SpaceObject.__init__(self, x, y, rng.random() * 2 * math.pi)

		self.size = size
		self.base_radius = 1
		self.num_vertices = 10 if self. size > SMALL else 8
		self.radius_variance = 0.2 * self.base_radius
		self.vel_x = rng.gauss(0, 3 * size / SMALL)
		self.vel_y = rng.gauss(0, 3 * size / SMALL)
		self.vel_angle = rng.gauss(0, .5)
		
		# Create model points/vertices
		self.radii = [
			round((self.base_radius + rng.gauss(0, 1) * self.radius_variance) * self.size)
			 for i in range(self.num_vertices)]
		self.model = []
		self.vertices = []
		for i, r in enumerate(self.radii):
			# create vertices from trig around a circle
			angle = 2 * math.pi * (i / self.num_vertices)
			px = math.cos(angle) * r
			py = math.sin(angle) * r
			self.model.append( (px, py) )
		self.update_model()
		
		# useful radii
		self.min_radius = min(self.radii)
		self.max_radius = max(self.radii)
		self.average_radius = sum(self.radii) / len(self.radii)
		
		# base collisions off of this radius:
		self.bounding_radius = (self.max_radius + self.average_radius) / 2

		# radius that the player collides with
		self.death_radius = (self.min_radius + self.average_radius) / 2
		
		# calculate mass with consistent density
		self.mass = math.pi * self.average_radius**2
		
		# deal with ids
		self.asteroid_id = self.get_next_asteroid_id()
	
	def update_model(self):
		pos_x, pos_y, angle = self.pos_x, self.pos_y, self.angle
		self.vertices = [util.transform_point(x, y, pos_x, pos_y, angle, 1, 1) for (x, y) in self.model]
	
	def update(self, screen):
		if self.alive:
			self.update_pos()
			self.update_model()
			util.draw_wrapped_lines(screen, YELLOW, True, self.vertices)

File: Java.old/Asteroid.java
package izak.asteroids;

import java.awt.Color;
import java.awt.Graphics2D;
import java.util.Random;

public class Asteroid extends SpaceObject {

	private static int lastAsteroidId = 0;
	private static final float DENSITY = 1;
	
	final Color darkBackground = new Color(15, 8, 8);
	
	final int asteroidId;
	final float mass;
	
	// radius and model information
	final float radiusVariance;
	final float baseRadius, averageRadius, minRadius, maxRadius, boundingRadius;
	final int numVertices;
	final Vector2F[] model;
	final float[] radii;
	Vector2F[] vertices;
	
	public Asteroid(float x, float y, float radius, Random random, int numVertices, float radiusVariance) {	
		// init stuffs
		this.x = x;
		this.y = y;
		this.angle = (float) (random.nextDouble() * 2 * Math.PI);
		this.baseRadius = radius;
		this.numVertices = numVertices;
		this.radiusVariance = radiusVariance;
		
		// init vertices
		model = new Vector2F[numVertices];
		vertices = new Vector2F[numVertices];
		radii = new float[numVertices];
		
		// populate model
		for (int i = 0; i < numVertices; i++) {
			// generate random radii
			float r = (float) (baseRadius + random.nextGaussian() * radiusVariance);
			radii[i] = r;
			// create vertices from trig around a circle
			double angle = 2 * Math.PI * ((double)i / (double)numVertices);
			float px = (float) (Math.cos(angle) * r);
			float py = (float) (Math.sin(angle) * r);
			model[i] = new Vector2F(px, py);
		}
		updateModel();
		
		averageRadius = Util.avg(radii);
		minRadius = Util.min(radii);
		maxRadius = Util.max(radii);
		
		// base collisions off of this radius
		boundingRadius = (maxRadius + averageRadius) / 2.0f;
		
		// calculate mass with consistent density
		mass = (float) (Math.PI * averageRadius*averageRadius * DENSITY);
		
		// deal with ids
		asteroidId = lastAsteroidId++;
	}
	
	private void updateModel() {
		// update model
		vertices = Util.transformPoints(model, x, y, angle, 1, 1);
	}
	
	public void update() {
		updatePos();
		updateModel();
	}
	
	public String toString() {
		return String.format("Asteroid:{x:%s, y:%s, r:%s, vx:%s, vy:%s, vr:%s, boundingRadius: %s, numVertices: %s", 
				x, y, angle, xVel, yVel, velAngle, boundingRadius, numVertices);
	}

	public Vector2F[] getCurrentModel() {
		return vertices;
	}
	
	public void draw(VectorGraphics vg) {
		Graphics2D g = vg.getGraphics();
//		// debug info
//		g.setColor(Color.white);
//		vg.drawEllipse(x, y, boundingRadius, boundingRadius);
//		vg.drawLine((int)(x - 3), (int)y, (int)(x+3), (int)y);
//		vg.drawLine((int)x, (int)(y - 3), (int)x, (int)(y+3));
		// draw red shape all around
		g.setColor(Color.yellow);
		vg.drawWrapLines(getCurrentModelInts(), true);
	}
}
